### DISCUSSION TITLE
Transitioning from accuracy coding to stimulus coding

### DISCUSSION THREAD

**Discussion by mberns-ru** [mwshinn/PyDDM#95] [Ask for help]
Hi! I have a model that accounts for a biased starting point, and all my functions worked fine when it was formatted for accuracy coding. We decided to change the bias model to the stimulus coding format because it is more intuitive with our experimental design. However, I've run into a few problems:

1. I can't extract some of the desired information on the fit. For example, the below line cannot be used on a sample with choices other than correct and incorrect:
`samp_mean = samp.mean_decision_time()`
I'm unsure how else to extract that information.

2. model.simulate_trial() produces the same trajectory across all conditions, even though the model fit produces a distribution that would imply different trajectories:
![image](https://github.com/mwshinn/PyDDM/assets/141958543/89bc37db-dd18-456c-bc6e-97c435704a23)
![image](https://github.com/mwshinn/PyDDM/assets/141958543/5ddc2d3b-e702-4d4f-a11a-41041790274c)
In the bottom graph, 4 distinct lines should be visible. I've played around with the model gui to confirm this.

The model I am using is this:
`model_fitted = Model(drift=DriftCoherence(drifttone=Fittable(minval=-20, maxval=20)),
                             noise=NoiseConstant(noise=1),
                             bound=BoundConstant(B=Fittable(minval=0.1, maxval=10)),
                             IC=ICPoint(x0=Fittable(minval=-0.1, maxval=0.1)),
                             overlay=OverlayChain(overlays=[OverlayNonDecision(nondectime=Fittable(minval=0, maxval=0.8))]),
                             T_dur = 2.5, dx=0.001, dt=0.001, choice_names=('Left', 'Right'))`

I'm unsure how to approach this. Please let me know if you have any suggestions!

**Comment by mwshinn** [comment]
1. The mean_decision_time function only looks at correct choices, but in stimulus coding, there is not necessarily a "correct" choice.  If you want the mean decision time of all choices in a Solution, you can compute it by hand using "np.sum((sol.choice_upper+sol.choice_lower)*self.t_domain) / (self.prob("upper")+self.prob("lower"))".  For a Sample you can just do it by averaging all your RTs.  (You don't need PyDDM for this.)
2. You have to set the seed for simulate_trial, it is zero by default.  Just pass the seed=x argument.  Also make sure to pass in different condition arguments if you are using a drift rate which varies based on a condition (as you are doing here).

Also, just as a heads up, you shouldn't normally need to use simulate_trial or simulated_solution.  Really the only reason to use these is to make a figure for a paper.

**Reply by mberns-ru** [reply]
Hey, thanks for the response!

1. I tried averaging the RTs, but I couldn't figure out how to access them with samp.items(choice) or samp.corr/samp.err (since they're not applicable). But you're right, I can just work around this with a pandas dataframe.
2. I have the seed set to 0 currently just for the sake of troubleshooting; in this case, I'm expecting 2 lines essentially mirrored across x0 (in other words, 2 different drift rates with the same seeded noise). But what's odd is that it plots the same line despite iterating through my different conditions:
```
for amrate in samp.condition_values('AMRate'):
        trajectory = model_fitted.simulate_trial(conditions={'AMRate': amrate}, seed=0)
        times = model_fitted.t_domain()
        times = times[:len(trajectory)]
        plt.plot(times, trajectory, label = str(round(amrate, 2)) + " kHz")
```
And I've used print statements to confirm that the AMRates are being called correctly, and I've also used the model gui to confirm that 2 am rates should generally drift left and the other 2 should drift right.

Also, yes; this plot is for paper purposes :-)

**Reply by mwshinn** [reply]
RE 2: Could you please post the full, minimal code necessary to reproduce the problem?  Something I can directly run and observe the problem.

**Reply by mberns-ru** [reply]
```
class DriftCoherence(ddm.models.Drift):
    name = "Drift depends linearly on AMRate distance from midpoint"
    required_parameters = ["drifttone"]
    required_conditions = ["AMRate"]

    def get_drift(self, conditions, **kwargs):
        return self.drifttone * (np.abs(np.log(conditions['AMRate']) - np.log(6.25)))
```
```
def graph_bias_trial(samp, title, graphs_folder):
    """
    Plot and save a biased simulated trial graph for each AMRate condition in the given sample.

    Parameters:
    samp (Sample): The sample object containing the data.
    title (str): The title of the graph.
    graphs_folder (str): The folder path where the graph will be saved.

    Returns:
    None
    """
    model_fitted = Model(drift=DriftCoherence(drifttone=Fittable(minval=-20, maxval=20)),
                             noise=NoiseConstant(noise=1),
                             bound=BoundConstant(B=Fittable(minval=0.1, maxval=10)),
                             IC=ICPoint(x0=Fittable(minval=-0.1, maxval=0.1)),
                             overlay=OverlayChain(overlays=[OverlayNonDecision(nondectime=Fittable(minval=0, maxval=0.8))]),
                             T_dur = 2.5, dx=0.001, dt=0.001, choice_names=('Left', 'Right'))
    model_fitted = fit_adjust_model(sample=samp, model=model_fitted, verbose=False, lossfunction=LossRobustLikelihood)
    ddm.plot.model_gui(model=model_fitted, sample=samp)
    for amrate in samp.condition_values('AMRate'):
        trajectory = model_fitted.simulate_trial(conditions={'AMRate': amrate}, seed=0)
        seed_count += 1
        times = model_fitted.t_domain()
        times = times[:len(trajectory)]
        plt.plot(times, trajectory, label = str(round(amrate, 2)) + " kHz")
    plt.show()
    return
```
My data has the following relevant columns:
- rt (float)
- choice_side (1 for left, 0 for right)
- AMRate (4.0, 5.0, 7.82, 9.77; the only condition)

**Reply by mwshinn** [reply]
Sorry for not being more clear, by "minimal code" and "something I can directly run" I meant something that really is minimal and really can be directly run, e.g., imports and the top and code stripped down to the minimum that reproduces the bug.  You shouldn't even need data for this or your function, just define the model with dummy parameters (no fittables) and then call simulate_trial with three example numbers.  Often by doing this you'll be able to find your bug without my help!
